### PR TITLE
Temperature anomaly postprocessor

### PR DESCRIPTION
**Pull request by maxrudolph**
Add a postprocessor for temperature - depth average of temperature. Also adds .update() function to visualization postprocessor classes.

### CODE REVIEW

File: source/postprocess/visualization/temperature_anomaly.cc [hunk 1]
```diff
@@ -0,0 +1,182 @@
...
+namespace aspect
+{
+  namespace Postprocess
+  {
+    namespace VisualizationPostprocessors
+    {
+      ASPECT_REGISTER_VISUALIZATION_POSTPROCESSOR(TemperatureAnomaly,
+                                                  "temperature anomaly",
+                                                  "A visualization output postprocessor that outputs the temperature minus the depth-average of the temperature."
+                                                  "The average temperature is calculated using the ``depth average'' postprocessor and interpolated linearly "
```
Comment (gassmoeller) [inline source/postprocess/visualization/temperature_anomaly.cc]: this sounds like the 'depth average' postprocessor has to be active in order to use this postprocessor, but that does not seem to be true. Would it be more correct to reword that it is calculated using the lateral averaging function that is also used by the depth average postprocessor?

**Comment by maxrudolph** [comment]
@gassmoeller do you have a preference of whether I add a brand new test or modify an existing test to add the new postprocessor?

**Comment by gassmoeller** [comment]
I would prefer a new test (the more separate tests we have the easier it is to pinpoint the problem in case they fail. You can copy one of the existing tests and add your postprocessor. Ideally use one, which outputs the visualization files as gnuplot files (no binary output), and stick to a small 2D model to keep the output small.  `visualization_conductivity_diffusivity` looks like a good candidate if you add a temperature anomaly to the vertical temperature profile of the test.

File: include/aspect/postprocess/visualization.h [hunk 2]
```diff
@@ -115,6 +115,8 @@ namespace aspect
            */
           virtual void initialize ();
 
+          virtual void update();
```
[the remark this answers is not in the document]
Comment (maxrudolph) [reply include/aspect/postprocess/visualization.h]: Done

[the remark this answers is not in the document]
Comment (maxrudolph) [reply include/aspect/postprocess/visualization.h]: Done.

File: include/aspect/postprocess/visualization/temperature_anomaly.h [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+          virtual
+          void
+          parse_parameters (ParameterHandler &prm);
+
+        private:
+          unsigned int n_slices;
+          double max_depth;
+          std::vector<double> padded_temperature_depth_average;
+          bool extrapolate_surface;
+          bool extrapolate_bottom;
```
[the remark this answers is not in the document]
Comment (maxrudolph) [reply include/aspect/postprocess/visualization/temperature_anomaly.h]: done.

File: source/postprocess/interface.cc [hunk 4]
```diff
@@ -42,33 +42,31 @@ namespace aspect
...
+    Interface<dim>::declare_parameters (ParameterHandler &)
+    {}
 
     template <int dim>
     void
     Interface<dim>::parse_parameters (ParameterHandler &)
     {}
 
-
-
```
[the remark this answers is not in the document]
Comment (maxrudolph) [reply source/postprocess/interface.cc]: Done.

File: source/postprocess/visualization/temperature_anomaly.cc (hunk 1 above)
Comment (maxrudolph) [reply source/postprocess/visualization/temperature_anomaly.cc]: Corrected.

File: source/postprocess/visualization/temperature_anomaly.cc [hunk 5]
```diff
@@ -0,0 +1,182 @@
...
+          {
+            padded_temperature_depth_average[n_slices+1] = 2.*temperature_depth_average[n_slices-1] - temperature_depth_average[n_slices-2];
+          }
+        else
+          {
+            const double bottom_temperature = this->get_boundary_temperature_manager().maximal_temperature(this->get_fixed_temperature_boundary_indicators());
+            padded_temperature_depth_average[n_slices+1] = 2.*bottom_temperature - temperature_depth_average[n_slices-1];
+          }
+        std::copy ( temperature_depth_average.begin(), temperature_depth_average.end(), padded_temperature_depth_average.begin() + 1 );
+        max_depth = this->get_geometry_model().maximal_depth();
```
[the remark this answers is not in the document]
Comment (maxrudolph) [reply source/postprocess/visualization/temperature_anomaly.cc]: OK, compromised to store it in a local variable

File: source/postprocess/visualization/temperature_anomaly.cc [hunk 6]
```diff
@@ -0,0 +1,182 @@
...
+      template <int dim>
+      TemperatureAnomaly<dim>::
+      TemperatureAnomaly ()
+        :
+        DataPostprocessorScalar<dim> ("temperature_anomaly",
+                                      update_values | update_q_points )
+      {
+        max_depth = 0.0;
+        std::vector<double> tmp(1);
+        padded_temperature_depth_average=tmp;
```
[the remark this answers is not in the document]
Comment (maxrudolph) [reply source/postprocess/visualization/temperature_anomaly.cc]: You're right, it shouldn't be necessary.

**Comment by maxrudolph** [comment]
@gassmoeller I think that I've addressed all of your comments.

File: include/aspect/postprocess/visualization/temperature_anomaly.h [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+
+          /**
+           * Read the parameters this class declares from the parameter file.
+           */
+          virtual
+          void
+          parse_parameters (ParameterHandler &prm);
+
+        private:
+          unsigned int n_slices; // number of slices to use when computing depth average of temperature
```
Comment (gassmoeller) [inline include/aspect/postprocess/visualization/temperature_anomaly.h]: Please make this a doxygen comment before the variable like the other one below. This way it shows up in the online documentation at https://aspect.geodynamics.org/doc/doxygen/index.html

File: include/aspect/postprocess/visualization/temperature_anomaly.h [hunk 8]
```diff
@@ -0,0 +1,88 @@
...
+          std::vector<double> padded_temperature_depth_average;
```
Comment (gassmoeller) [inline include/aspect/postprocess/visualization/temperature_anomaly.h]: Please add a comment for this variable as well, it is the one that is most confusing without explanation.

**Comment by maxrudolph** [comment]
@gassmoeller not sure if you looked at the tester output but it appears that there must be a bug in the plugin still as there was a real error. I will try to sort this out.

**Comment by gassmoeller** [comment]
No, I did not check, I just saw that the testers were failing. It looks like the test result is fine now, but there is an indentation problem in the assertion (I think you have to replace a tab by spaces).

**Comment by maxrudolph** [comment]
OK, I think that everything has been corrected now.

**Comment by gassmoeller** [comment]
Looks good. Thanks!

**Pull request merged**